akka-typed/src/main/scala/akka/typed/persistence/scaladsl/PersistentActor.scala

```diff
@@ -20,20 +20,29 @@ object PersistentActor {
       recoveryCompleted = (state, _) ⇒ state)
 
   sealed abstract class PersistentEffect[+Event, State]() {
-    def andThen(callback: State ⇒ Unit): PersistentEffect[Event, State]
-  }
 
-  final case class PersistNothing[Event, State](callbacks: List[State ⇒ Unit] = Nil) extends PersistentEffect[Event, State] {
-    def andThen(callback: State ⇒ Unit) = copy(callbacks = callback :: callbacks)
+    def andThen(sideEffect: ChainableEffect[_, State]): PersistentEffect[Event, State] =
+      CompositeEffect(this, List(sideEffect))
+
+    /** Convenience method to register a side effect with just a callback function */
+    def andThen(callback: State ⇒ Unit): PersistentEffect[Event, State] = andThen(SideEffect(callback))
   }
 
-  case class Persist[Event, State](event: Event, callbacks: List[State ⇒ Unit] = Nil) extends PersistentEffect[Event, State] {
-    def andThen(callback: State ⇒ Unit) = copy(callbacks = callback :: callbacks)
+  case class CompositeEffect[Event, State](mainEffect: PersistentEffect[Event, State], sideEffects: List[ChainableEffect[_, State]]) extends PersistentEffect[Event, State] {
+    override def andThen(sideEffect: ChainableEffect[_, State]): PersistentEffect[Event, State] =
+      copy(sideEffects = sideEffects :+ sideEffect)
   }
 
-  case class Unhandled[Event, State](callbacks: List[State ⇒ Unit] = Nil) extends PersistentEffect[Event, State] {
-    def andThen(callback: State ⇒ Unit) = copy(callbacks = callback :: callbacks)
+  case class PersistNothing[Event, State]() extends PersistentEffect[Event, State]
+  case class Persist[Event, State](event: Event) extends PersistentEffect[Event, State]
+
+  trait ChainableEffect[Event, State] {
```

Comment:
do we need to make a difference?
Persist(evt1).andThen(Persist(evt2)) is not impossible, or wrong

Reply:
My thinking here was that it might help to separate the `Persist` (which depends on `[Event]` but is independent of `[State]`) from the `SideEffect` (which depends on `[State]` but is independent of `[Event]`) might help keeping the types inferrable. 

Also, do we want to allow `Persist(evt1).andThen(SideEffect(effect1)).andThen(Persist(evt2))`? That might make the implementation a bit harder, and I'm not sure it's something we'd want to promote.

Reply:
I don't see any bad consequences of allowing such chaining. Isn't it like a depth first traversal of a tree structure? We don't expect any deep structures so don't have to worry about recursion depth.

Reply:
It does mean making 2 database calls (in the mean time 'holding up' processing new commands), but indeed it might not be a problem